File: cli-rust/crates/oa-cmd-auth/src/api_key.rs
/// API key normalization, validation, and preview formatting.
///
/// Handles shell-style assignments (`export KEY="value"`), unquoting,
/// trailing semicolons, and produces safe preview strings for user display.
///
/// Source: `src/commands/auth-choice.api-key.ts`

use regex::Regex;
use std::sync::LazyLock;

/// Default number of characters shown at the head of a key preview.
///
/// Source: `src/commands/auth-choice.api-key.ts` - `DEFAULT_KEY_PREVIEW`
const DEFAULT_KEY_PREVIEW_HEAD: usize = 4;

/// Default number of characters shown at the tail of a key preview.
///
/// Source: `src/commands/auth-choice.api-key.ts` - `DEFAULT_KEY_PREVIEW`
const DEFAULT_KEY_PREVIEW_TAIL: usize = 4;

/// Regex matching shell-style assignments like `export KEY="value"` or `KEY=value`.
///
/// Source: `src/commands/auth-choice.api-key.ts` - `normalizeApiKeyInput`
static ASSIGNMENT_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^(?:export\s+)?[A-Za-z_][A-Za-z0-9_]*\s*=\s*(.+)$")
        .expect("ASSIGNMENT_RE is a valid regex")
});

/// Normalize raw API key input.
///
/// Strips shell-style `export FOO="bar"` wrappers, removes surrounding
/// quotes (single, double, or backtick), and trims trailing semicolons.
///
/// # Examples
///
/// ```
/// use oa_cmd_auth::api_key::normalize_api_key_input;
///
/// assert_eq!(normalize_api_key_input("sk-abc123"), "sk-abc123");
/// assert_eq!(normalize_api_key_input("export KEY=\"sk-abc123\""), "sk-abc123");
/// assert_eq!(normalize_api_key_input("KEY='sk-abc123';"), "sk-abc123");
/// ```
///
/// Source: `src/commands/auth-choice.api-key.ts` - `normalizeApiKeyInput`
pub fn normalize_api_key_input(raw: &str) -> String {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return String::new();
    }

    // Handle shell-style assignments: export KEY="value" or KEY=value
    let value_part = ASSIGNMENT_RE
        .captures(trimmed)
        .and_then(|caps| caps.get(1))
        .map_or(trimmed, |m| m.as_str().trim());

    // Strip trailing semicolon first (handles `KEY='val';` pattern)
    let without_semi = value_part.strip_suffix(';').unwrap_or(value_part);

    // Unquote surrounding quotes (double, single, or backtick)
    let unquoted = if without_semi.len() >= 2 {
        let bytes = without_semi.as_bytes();
        let first = bytes[0];
        let last = bytes[bytes.len() - 1];
        if (first == b'"' && last == b'"')
            || (first == b'\'' && last == b'\'')
            || (first == b'`' && last == b'`')
        {
            &without_semi[1..without_semi.len() - 1]
        } else {
            without_semi
        }
    } else {
        without_semi
    };

    // Strip any remaining trailing semicolon (in case it was inside quotes)
    let final_str = unquoted.strip_suffix(';').unwrap_or(unquoted);

    final_str.trim().to_owned()
}

/// Validate an API key input string.
///
/// Returns `None` if the input is valid (non-empty after normalization),
/// or `Some("Required")` if it is empty.
///
/// Source: `src/commands/auth-choice.api-key.ts` - `validateApiKeyInput`
#[must_use]
pub fn validate_api_key_input(value: &str) -> Option<&'static str> {
    if normalize_api_key_input(value).is_empty() {
        Some("Required")
    } else {
        None
    }
}

/// Options for controlling how much of a key is shown in a preview.
///
/// Source: `src/commands/auth-choice.api-key.ts` - `formatApiKeyPreview` opts
#[derive(Debug, Clone, Copy)]
pub struct KeyPreviewOpts {
    /// Characters to show at the beginning.
    pub head: usize,
    /// Characters to show at the end.
    pub tail: usize,
}

impl Default for KeyPreviewOpts {
    fn default() -> Self {
        Self {
            head: DEFAULT_KEY_PREVIEW_HEAD,
            tail: DEFAULT_KEY_PREVIEW_TAIL,
        }
    }
}

/// Format an API key as a safe preview string for user display.
///
/// Shows the first `head` and last `tail` characters separated by an
/// ellipsis. Short keys use a minimal preview to avoid revealing the
/// full value.
///
/// # Examples
///
/// ```
/// use oa_cmd_auth::api_key::format_api_key_preview;
///
/// assert_eq!(format_api_key_preview("sk-abcdef1234567890", None), "sk-a\u{2026}7890");
/// assert_eq!(format_api_key_preview("", None), "\u{2026}");
/// ```
///
/// Source: `src/commands/auth-choice.api-key.ts` - `formatApiKeyPreview`
#[must_use]
pub fn format_api_key_preview(raw: &str, opts: Option<KeyPreviewOpts>) -> String {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return "\u{2026}".to_owned();
    }

    let opts = opts.unwrap_or_default();
    let head = opts.head;
    let tail = opts.tail;

    if trimmed.len() <= head + tail {
        // Short key: show minimal preview
        let short_head = std::cmp::min(2, trimmed.len());
        let short_tail = std::cmp::min(2, trimmed.len().saturating_sub(short_head));
        if short_tail == 0 {
            return format!("{}\u{2026}", &trimmed[..short_head]);
        }
        return format!(
            "{}\u{2026}{}",
            &trimmed[..short_head],
            &trimmed[trimmed.len() - short_tail..]
        );
    }

    format!(
        "{}\u{2026}{}",
        &trimmed[..head],
        &trimmed[trimmed.len() - tail..]
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    // ── normalize_api_key_input ──

    #[test]
    fn normalize_empty_string() {
        assert_eq!(normalize_api_key_input(""), "");
        assert_eq!(normalize_api_key_input("   "), "");
    }

    #[test]
    fn normalize_plain_key() {
        assert_eq!(normalize_api_key_input("sk-abc123"), "sk-abc123");
    }

    #[test]
    fn normalize_with_whitespace() {
        assert_eq!(normalize_api_key_input("  sk-abc123  "), "sk-abc123");
    }

    #[test]
    fn normalize_export_double_quoted() {
        assert_eq!(
            normalize_api_key_input(r#"export API_KEY="sk-abc123""#),
            "sk-abc123"
        );
    }

    #[test]
    fn normalize_export_single_quoted() {
        assert_eq!(
            normalize_api_key_input("export API_KEY='sk-abc123'"),
            "sk-abc123"
        );
    }

    #[test]
    fn normalize_assignment_no_export() {
        assert_eq!(
            normalize_api_key_input("MY_KEY=sk-abc123"),
            "sk-abc123"
        );
    }

    #[test]
    fn normalize_trailing_semicolon() {
        assert_eq!(
            normalize_api_key_input("export KEY='sk-abc123';"),
            "sk-abc123"
        );
    }

    #[test]
    fn normalize_backtick_quoted() {
        assert_eq!(
            normalize_api_key_input("KEY=`sk-abc123`"),
            "sk-abc123"
        );
    }

    // ── validate_api_key_input ──

    #[test]
    fn validate_empty_is_required() {
        assert_eq!(validate_api_key_input(""), Some("Required"));
        assert_eq!(validate_api_key_input("   "), Some("Required"));
    }

    #[test]
    fn validate_nonempty_is_ok() {
        assert_eq!(validate_api_key_input("sk-abc"), None);
    }

    // ── format_api_key_preview ──

    #[test]
    fn preview_empty_string() {
        assert_eq!(format_api_key_preview("", None), "\u{2026}");
    }

    #[test]
    fn preview_long_key() {
        assert_eq!(
            format_api_key_preview("sk-abcdef1234567890", None),
            "sk-a\u{2026}7890"
        );
    }

    #[test]
    fn preview_short_key() {
        // "abc" has 3 chars, head+tail=8, so it uses the short branch
        let result = format_api_key_preview("abc", None);
        assert_eq!(result, "ab\u{2026}c");
    }

    #[test]
    fn preview_very_short_key() {
        let result = format_api_key_preview("a", None);
        assert_eq!(result, "a\u{2026}");
    }

    #[test]
    fn preview_custom_opts() {
        let opts = KeyPreviewOpts { head: 2, tail: 2 };
        assert_eq!(
            format_api_key_preview("sk-abcdef1234567890", Some(opts)),
            "sk\u{2026}90"
        );
    }
}

File: backend/internal/goproviders/authchoice/api_key.go
// authchoice/api_key.go — API Key 输入规范化工具
// 对应 TS 文件: src/commands/auth-choice.api-key.ts
package authchoice

import "strings"

// defaultKeyPreviewHead 预览显示的头部字符数（默认值）。
const defaultKeyPreviewHead = 4

// defaultKeyPreviewTail 预览显示的尾部字符数（默认值）。
const defaultKeyPreviewTail = 4

// NormalizeApiKeyInput 规范化 API Key 输入。
// 处理 shell 风格赋值语句（如 export KEY="value"）、引号和尾部分号。
// 对应 TS: normalizeApiKeyInput()
func NormalizeApiKeyInput(raw string) string {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return ""
	}

	// 处理 shell 风格赋值: export KEY="value" 或 KEY=value
	valuePart := trimmed
	// 匹配 export? IDENTIFIER = value 模式
	if idx := findAssignmentValue(trimmed); idx >= 0 {
		valuePart = strings.TrimSpace(trimmed[idx:])
	}

	// 去除引号
	unquoted := stripQuotes(valuePart)

	// 去除尾部分号
	if strings.HasSuffix(unquoted, ";") {
		unquoted = unquoted[:len(unquoted)-1]
	}

	return strings.TrimSpace(unquoted)
}

// findAssignmentValue 查找赋值语句中值部分的起始索引。
// 返回 '=' 后第一个非空白字符的索引，如果不是赋值语句则返回 -1。
func findAssignmentValue(s string) int {
	rest := s
	// 跳过可选的 "export "
	if strings.HasPrefix(rest, "export") && len(rest) > 6 && (rest[6] == ' ' || rest[6] == '\t') {
		rest = strings.TrimSpace(rest[6:])
	}
	// 检查标识符部分
	i := 0
	if i < len(rest) && isIdentStart(rest[i]) {
		i++
		for i < len(rest) && isIdentContinue(rest[i]) {
			i++
		}
		// 跳过空白到 '='
		j := i
		for j < len(rest) && (rest[j] == ' ' || rest[j] == '\t') {
			j++
		}
		if j < len(rest) && rest[j] == '=' {
			offset := len(s) - len(rest)
			return offset + j + 1
		}
	}
	return -1
}

func isIdentStart(c byte) bool {
	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '_'
}

func isIdentContinue(c byte) bool {
	return isIdentStart(c) || (c >= '0' && c <= '9')
}

// stripQuotes 去除首尾成对的引号（双引号、单引号或反引号）。
func stripQuotes(s string) string {
	if len(s) >= 2 {
		first, last := s[0], s[len(s)-1]
		if (first == '"' && last == '"') ||
			(first == '\'' && last == '\'') ||
			(first == '`' && last == '`') {
			return s[1 : len(s)-1]
		}
	}
	return s
}

// ValidateApiKeyInput 验证 API Key 输入（规范化后非空即有效）。
// 对应 TS: validateApiKeyInput
func ValidateApiKeyInput(value string) string {
	if NormalizeApiKeyInput(value) != "" {
		return ""
	}
	return "Required"
}

// FormatApiKeyPreview 格式化 API Key 预览显示。
// 显示头部和尾部部分字符，中间用 "…" 替代。
// 对应 TS: formatApiKeyPreview()
func FormatApiKeyPreview(raw string, head, tail int) string {
	trimmed := strings.TrimSpace(raw)
	if trimmed == "" {
		return "…"
	}
	if head <= 0 {
		head = defaultKeyPreviewHead
	}
	if tail <= 0 {
		tail = defaultKeyPreviewTail
	}
	if len(trimmed) <= head+tail {
		shortHead := min(2, len(trimmed))
		shortTail := min(2, len(trimmed)-shortHead)
		if shortTail <= 0 {
			return trimmed[:shortHead] + "…"
		}
		return trimmed[:shortHead] + "…" + trimmed[len(trimmed)-shortTail:]
	}
	return trimmed[:head] + "…" + trimmed[len(trimmed)-tail:]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
